File: src/solver/static_solver.cc
/**
 * @file   static_solver.cc
 * @author Aurelia Cuba Ramos <aurelia.cubaramos@epfl.ch>
 * @date   Wed Jul 30 15:35:01 2014
 *
 * @brief  implementation of the static solver
 *
 * @section LICENSE
 *
 * Copyright (©) 2010-2011 EPFL (Ecole Polytechnique Fédérale de Lausanne)
 * Laboratory (LSMS - Laboratoire de Simulation en Mécanique des Solides)
 *
 * Akantu is free  software: you can redistribute it and/or  modify it under the
 * terms  of the  GNU Lesser  General Public  License as  published by  the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Akantu is  distributed in the  hope that it  will be useful, but  WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A  PARTICULAR PURPOSE. See  the GNU  Lesser General  Public License  for more
 * details.
 *
 * You should  have received  a copy  of the GNU  Lesser General  Public License
 * along with Akantu. If not, see <http://www.gnu.org/licenses/>.
 *
 */

/* -------------------------------------------------------------------------- */
#include "static_solver.hh"

#if defined(AKANTU_USE_MPI)
#  include "mpi_type_wrapper.hh"
#endif

/* -------------------------------------------------------------------------- */
#ifdef AKANTU_USE_PETSC
#include <petscsys.h>
#endif

/* -------------------------------------------------------------------------- */

__BEGIN_AKANTU__

/* -------------------------------------------------------------------------- */
StaticSolver::StaticSolver() : CommunicatorEventHandler(), is_initialized(false) {
  StaticCommunicator::getStaticCommunicator().registerEventHandler(*this);
}


/* -------------------------------------------------------------------------- */
StaticSolver::~StaticSolver() {
  --this->nb_references;
  if(this->nb_references == 0) {
    StaticCommunicator::getStaticCommunicator().unregisterEventHandler(*this);
    delete this->static_solver;
  }
}

/* -------------------------------------------------------------------------- */
StaticSolver & StaticSolver::getStaticSolver() {
  if(nb_references == 0)
    static_solver = new StaticSolver();
  ++nb_references;
  return *static_solver;
}

#ifdef AKANTU_USE_PETSC
#endif

/* -------------------------------------------------------------------------- */
void StaticSolver::initialize(int & argc, char ** & argv) {
  if (this->is_initialized) return;
  //  AKANTU_DEBUG_ASSERT(this->is_initialized != true, "The static solver has already been initialized");


  this->is_initialized = true;
}

/* -------------------------------------------------------------------------- */
void StaticSolver::finalize() {
  ParentEventHandler::sendEvent(StaticSolverEvent::BeforeStaticSolverDestroyEvent());


  AKANTU_DEBUG_ASSERT(this->is_initialized == true, "The static solver has not been initialized");
#ifdef AKANTU_USE_PETSC
  PetscFinalize();
#endif

  this->is_initialized = false;
}

__END_AKANTU__

File: test/python_fe/static_solver.py
# ------------------------------------------------------------------------------
__author__ = "Nicolas Richart"
__copyright__ = "Copyright (C) 2016-2018, EPFL (Ecole Polytechnique Fédérale" \
                " de Lausanne) Laboratory (LSMS - Laboratoire de Simulation" \
                " en Mécanique des Solides)"
__credits__ = ["Nicolas Richart"]
__license__ = "L-GPLv3"
__maintainer__ = "Nicolas Richart"
__email__ = "nicolas.richart@epfl.ch"
# ------------------------------------------------------------------------------

import copy
import scipy.sparse.linalg as spl
from . import export
from . import fe


@export
class StaticSolver(fe.Solver):
    def __init__(self, model):
        self._model = model

    def _assembleResidual(self):
        self._r = self._model.f_ext - self._model.f_int

    def _assembleJacobian(self):
        self._J = copy.copy(self._model.K)
        self._model.applyDirichletBC()

        self._zero_rows(self._J, self._model.blocked)

    def solveStep(self):
        self._assembleJacobian()
        self._assembleResidual()
        x = spl.spsolve(self._J, self._r)
        self._model.u += x
